switch-to-configuration: bump deps and add more logging

jmbaur opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [Nixpkgs 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/doc/release-notes/rl-2511.section.md) (or backporting [25.05](https://github.com/NixOS/nixpkgs/blob/master/doc/manual/release-notes/rl-2505.section.md) Nixpkgs Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
- [NixOS 25.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2511.section.md) (or backporting [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) NixOS Release notes)
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md), [pkgs/README.md](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md), [maintainers/README.md](https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md) and other contributing documentation in corresponding paths.

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

jmbaur committed:
switch-to-configuration-ng: update dependencies

jmbaur committed:
switch-to-configuration-ng: add more logging

Adds more logging to the non-systemd related parts of
switch-to-configuration.

jmbaur commented:
@ofborg test switchTest

dtomvan commented on pkgs/by-name/sw/switch-to-configuration-ng/src/src/main.rs:
```diff
@@ -123,7 +123,14 @@ impl From<&Action> for &'static str {
 // Perl's "die" uses errno to set the exit code: https://perldoc.perl.org/perlvar#%24%21
 fn die() -> ! {
-    std::process::exit(std::io::Error::last_os_error().raw_os_error().unwrap_or(1));
+    let code = match std::io::Error::last_os_error().raw_os_error().unwrap_or(1) {
+        // Ensure that even if errno did not point to a helpful error code, we still have a
+        // non-zero exit code
+        0 => 1,
+        other => other,
+    };
+
+    std::process::exit(code);
```
This is more of a fix. You should call it that I think

dtomvan commented on pkgs/by-name/sw/switch-to-configuration-ng/src/src/main.rs:
```diff
@@ -988,15 +995,23 @@ dry-activate: show what would be done if this configuration were activated
     let install_bootloader = required_env("INSTALL_BOOTLOADER")?;
     let locale_archive = required_env("LOCALE_ARCHIVE")?;
     let new_systemd = PathBuf::from(required_env("SYSTEMD")?);
+    let log_level = if std::env::var("STC_DEBUG").is_ok() {
```
Is `STC_DEBUG` documented?

jmbaur replied:
Added documentation

dtomvan commented on pkgs/by-name/sw/switch-to-configuration-ng/src/src/main.rs:
```diff
@@ -1115,6 +1134,7 @@ won't take effect until you reboot the system.
     let mut units_to_reload = map_from_list_file(RELOAD_LIST_FILE);
 
     let dbus_conn = LocalConnection::new_system().context("Failed to open dbus connection")?;
+
```
Unrelated formatting change?

dtomvan commented on pkgs/by-name/sw/switch-to-configuration-ng/src/src/main.rs:
```diff
@@ -1691,6 +1711,7 @@ won't take effect until you reboot the system.
                     .canonicalize()
                     .context("Failed to get full path to /proc/self/exe")?;
 
+                log::debug!("Performing user switch for {}", name);
```
Nit

```suggestion
                log::debug!("Performing user switch for {name}");
```

dtomvan commented on pkgs/by-name/sw/switch-to-configuration-ng/src/src/main.rs:
```diff
@@ -1037,12 +1053,13 @@ fn do_system_switch(action: Action) -> anyhow::Result<()> {
         die();
     };
 
+    log::debug!("Acquiring lock on file /run/nixos/switch-to-configuration.lock");
```
Might be worth it to factor `/run/nixos/switch-to-configuration.lock` out to a constant.

dtomvan commented on pkgs/by-name/sw/switch-to-configuration-ng/src/Cargo.toml:
These updates feel a bit arbitrary. `anyhow`, `glob`, `log`, `regex` and `rust-ini` also have had updates, while the ones you did update look insignificant, except maybe for `nix`. `rust-ini` could be a tagged release again too if we update. Any reason why you did these specific updates?

jmbaur replied:
I used `cargo update --verbose` to update non-patch versions of dependencies, looks like the others you mentioned do not have non-patch versions that differ with our Cargo.toml here, thus they have the updated patch versions specified in Cargo.lock.

dtomvan replied:
I see. I don't understand why cargo accepts a patch version if it will not be pinned to it, but that doesn't matter right now. Well, could you maybe un-git-dep `rust-ini`?

jmbaur replied:
Because of semver, presumably crates of the same patch version will remain compatible

jmbaur replied:
Yeah I'll bump rust-ini, looks like there was a release 3 days ago that will allow for us to not use the git dep

dtomvan reviewed (commented):
Changes LGTM, I do have some nits here and there but generally it looks fine.

Would appreciate it though if you split the change off where you fixed `die()`

jmbaur committed:
switch-to-configuration-ng: ensure die() always exits with non-zero code

dtomvan commented on pkgs/by-name/sw/switch-to-configuration-ng/src/src/main.rs:
```diff
@@ -1866,10 +1895,17 @@ won't take effect until you reboot the system.
-    {}
+    let mut waited = Duration::from_millis(0);
+    let wait_interval = Duration::from_millis(250);
+    let max_wait = Duration::from_secs(90);
```
I see you introduced a timeout, which probably also warrants a seperate commit.

dtomvan reviewed (approved):
Diff LGTM. 1 nit

jmbaur committed:
switch-to-configuration-ng: add timeout while waiting for settle

Without a timeout, we (switch-to-configuration) could be stuck in a loop
if we continue receiving D-Bus signals for new systemd events. This
could indicate a problem elsewhere on the machine, however we shouldn't
allow this to cause switch-to-configuration to get stuck.
